pkg/services/caching/service.go

```diff
@@ -0,0 +1,61 @@
+package caching
+
+import (
+	"context"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+)
+
+const (
+	XCacheHeader   = "X-Cache"
+	StatusHit      = "HIT"
+	StatusMiss     = "MISS"
+	StatusBypass   = "BYPASS"
+	StatusError    = "ERROR"
+	StatusDisabled = "DISABLED"
+)
+
+type CacheQueryResponseFn func(context.Context, *backend.QueryDataResponse)
+
+type CachedQueryDataResponse struct {
+	// The cached data response associated with a query, or nil if no cached data is found
+	Response *backend.QueryDataResponse
+	// A function that should be used to cache a QueryDataResponse for a given query.
+	// It can be set to nil by the method implementation (if there is an error, for example), so it should be checked before being called.
+	UpdateCacheFn CacheQueryResponseFn
+}
+
+func ProvideCachingService() *OSSCachingService {
+	return &OSSCachingService{}
+}
+
+//go:generate mockery --name CachingService --structname FakeOSSCachingService --inpackage --filename caching_service_mock.go
+type CachingService interface {
```

Comment:
Received a question on this in slack that I'll address here for others:

Question: Why does func (m *CachingMiddleware) QueryData update the cache by calling cr.UpdateCacheFn(ctx, resp) but func (m *CachingMiddleware) CallResource expects the caller to update the cache?

Answer: 
There are two reasons it’s done this way, one is a design choice, and the other is because it’s kind of forced on us due to the plugin api.

Design choice: when the query cache handler func is called, we do a lot of work up front to validate the queries, retrieve the cache config, generate a key, etc. I thought it would be inefficient to do all that stuff twice, once for the cache retrieval, and another for the update. By returning a function, I can just do it all once and also give us better control over calls to update the cache in general.

Plugin API: there’s also a major difference in how the query and resource calls work. With the query call, you get back a single query response synchronously, so getting just that result and caching it is straightforward. With the resource API, there can be multiple responses, and they are sent from the plugin via a stream, which we read higher up the call hierarchy. If I wanted to return a cache func for the resource call, I would either have to read from the stream in the middleware, or pass that update func way back up the call stack - neither of these are very easily done using the plugin middleware

Reply:
Thanks. Had similar questions. It's unfortunate the call resource streaming logic lives in the API layer. It's quite a code smell. I think the following logic could be moved downwards to plugin client:
https://github.com/grafana/grafana/blob/1c3ce0735f3b6e5609c3a7d8988d7f0700ee75b1/pkg/api/plugin_resource.go#L167-L192

Plugin client already does some cleaning of HTTP response headers:
https://github.com/grafana/grafana/blob/1c3ce0735f3b6e5609c3a7d8988d7f0700ee75b1/pkg/plugins/manager/client/client.go#L96-L103

In call resource handle of caching client middleware you can implement similar `cachedSender` as above and do similar things with caching you're currently doing in plugin_resource.go

With that said, I think it could be a nice follow up task for the future.

Reply:
And in addition, moving things downstream is probably a task for @grafana/plugins-platform-backend to pick up eventually